src/libNetwork/P2PComm.cpp

```diff
@@ -387,49 +391,66 @@ void P2PComm::HandleAcceptedConnection(
     if (buf[1] == START_BYTE_BROADCAST)
     {
         read_length = 0;
-        while (read_length != HASH_LEN)
+
         {
-            int n = read(cli_sock, hash_buf + read_length,
-                         HASH_LEN - read_length);
-            if (n <= 0)
+            lock_guard<mutex> guard(
+                P2PComm::GetInstance().m_receiveMessageMutex);
+
+            while (read_length != HASH_LEN)
             {
-                LOG_GENERAL(WARNING,
-                            "Socket read failed. Code = "
-                                << errno << " Desc: " << std::strerror(errno)
-                                << ". IP address: " << from);
-                return;
+                int n = read(cli_sock, hash_buf + read_length,
+                             HASH_LEN - read_length);
+                if (n <= 0)
+                {
+                    LOG_GENERAL(WARNING,
+                                "Socket read failed. Code = "
+                                    << errno
+                                    << " Desc: " << std::strerror(errno)
+                                    << ". IP address: " << from);
+                    return;
+                }
+                read_length += n;
             }
-            read_length += n;
         }
 
         // Check if this message has been received before
         bool found = false;
         {
-            lock_guard<mutex> guard(
-                P2PComm::GetInstance().m_broadcastHashesMutex);
             vector<unsigned char> msg_hash(hash_buf, hash_buf + HASH_LEN);
-            found = (P2PComm::GetInstance().m_broadcastHashes.find(msg_hash)
-                     != P2PComm::GetInstance().m_broadcastHashes.end());
+
+            {
+                lock_guard<mutex> guard(
```

Comment:
Better make methods in `P2PComm` instead of using locks outside. (the original code also had the same issue)